MobileManager/Services/AppResourceService.cs

```diff
@@ -0,0 +1,154 @@
+    {
+        private readonly IManagerLogger _logger;
+        private readonly IRestClient _restClient;
+        private readonly IRepository<AppResourceInfo> _appResourceRepository;
+        private readonly string _storagePath;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="T:MobileManager.Services.AppResourceService"/> class.
+        /// </summary>
+        /// <param name="restClient">Rest client.</param>
+        /// <param name="appResourceRepository">appResourceRepository.</param>
+        /// <param name="logger">Logger.</param>
+        public AppResourceService(IRestClient restClient, IRepository<AppResourceInfo> appResourceRepository, IManagerLogger logger) : base(logger)
+        {
+            _logger = logger;
+            _restClient = restClient;
+            _appResourceRepository = appResourceRepository;
+            _storagePath = Path.Combine(Directory.GetCurrentDirectory(), "../_appFiles");
+        }
+
+        /// <inheritdoc />
+        public AppResourceInfo FindAppResource(string id)
+        {
+            return _appResourceRepository.Find(id);
+        }
+
+        /// <inheritdoc />
+        public bool DeleteAppResource(string id)
+        {
+            var appFolder = Path.Combine(_storagePath, id);
+            var result = _appResourceRepository.Remove(id);
+            Directory.Delete(appFolder, true);
+            return true;
+        }
+
+        /// <inheritdoc />
+        public async Task<AppResourceInfo> UploadAppResource(IFormFile file)
+        {
+            CreateStorageFolder();
+
+            var extension = Path.GetExtension(file.FileName).Trim('.').ToLower();
+            var hash = GetHashFromFile(file.OpenReadStream());
+            var folderPath = Path.Combine(_storagePath, hash);
+            var filePath = Path.Combine(folderPath, file.FileName);
+
+            if (!Directory.Exists(folderPath))
+            {
+                Directory.CreateDirectory(folderPath);
+                var newStream = new FileStream(filePath, FileMode.Create);
+                await file.CopyToAsync(newStream);
+                var appInfo = new AppResourceInfo(hash, file.FileName, filePath, extension, DateTime.Now);
+                _appResourceRepository.Add(appInfo);
+                newStream.Close();
+                return appInfo;
+            }
+
+            return FindAppResource(hash);
+        }
+
+        /// <inheritdoc />
+        public AppResourceInfo DownloadAppResourceFromUri(Uri uri)
+        {
+            /* 
+            * at first try HEAD HTTP request
+            * if is MD5 there we can use it without necessity to download whole file 
+            * It works for e.g. artifactory 
+            */
+            WebRequest request = WebRequest.Create(uri);
+            request.Method = "HEAD";
+            var md5 = request.GetResponse().Headers.Get("X-Checksum-Md5");

```

Comment:
you have no exception handling here - it might timeout or incorrect url and it will throw nasty exception back to the user - should be probably handled by BadRequest 

and no handling in the controller also - so maybe put it to the controller

Reply:
i will add exception handling into controller